atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManagerImpl.java

```diff
@@ -114,6 +115,34 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
         return numTimesFailed > NUM_RETRIES;
     }
 
+    public <C extends PreCommitCondition, E extends Exception> ServiceWriteTransaction getWriteTransaction(C condition)
+            throws E {
+        checkOpen();
+        try {
+            RawTransaction tx = setupRunTaskWithConditionThrowOnConflict(condition);
+            return new ServiceWriteTransaction(tx.delegate()) {
+                @Override
+                public void commit() {
+                    tx.commit();
+                    finishRunTaskWithLockThrowOnConflict(tx, transaction -> null);
+                }
+            };
+        } finally {
+            condition.cleanup();
```

Comment:
The condition needs to be cleaned up post-commit, just like `finishRunTaskWithLockThrowOnConflict`. This will clean it up immediately after creating the transaction.